### PR TITLE
[WIP] EEG-Inception

### PR DESCRIPTION
**Pull request by bruAristimunha**
I am proposing to implement EEG-Inception. This model was used by the winners of last year [BEETL Competition](https://arxiv.org/pdf/2202.03267.pdf).

Is this suitable for library?

### CODE REVIEW

File: braindecode/models/eeginception.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    ):
+        super().__init__()
+
+        self.n_channels = n_channels
+        self.n_classes = n_classes
+        self.input_size_s = input_size_s
+        self.drop_prob = drop_prob
+        self.sfreq = sfreq
+
+        input_samples = int(input_size_s * sfreq / 1000)
```
Comment (robintibor) [inline braindecode/models/eeginception.py:103]: seems input samples not used

**Comment by robintibor** [comment]
In general more models are very suitable thank you very much!! Can you add some tests with different input sizes just checking forward works and output shape is as expected?

**Comment by agramfort** [comment]
great stuff as @robintibor would say :)

to get this merged we would need a test (see how done for others) and a what's new entry

btw I checked the license of https://github.com/Grifcc/EEG/tree/90e412a407c5242dfc953d5ffb490bdb32faf022 and it's Apache2 so it's ok for here (BSD).

**Comment by robintibor** [comment]
Do you need any support here? :) Is this still ongoing? :)

**Comment by bruAristimunha** [comment]
thanks for reminding me @robintibor, I will try to finish this week.

**Comment by bruAristimunha** [comment]
Hallo @robintibor!

I think I'm going to need some help X= 
It is a little more difficult than expected to adjust the dimensions of the model.

**Comment by bruAristimunha** [comment]
Hello @sliwy, how are you?

I'm also in need of code review. I'm doing something wrong in the code and it's not working for dimensions other than the default value.

**Comment by sliwy** [comment]
Hi @bruAristimunha , I may try to help. I need more info from you when it fails, what is the error, maybe how to reproduce. 

I think that we may be able to simplify a little bit by creating blocks and using `nn.Sequential` in the `EEGInception` implementation as well. I'm going to put some comments in the review.

**Comment by bruAristimunha** [comment]
Hi @sliwy,

Strangely, the change to sequential increases the number of parameters. It still has a weird error when changing input values. I re-added the tests that were failing. I have no idea what's wrong or how to fix. I think I'll wait for the conclusion of the other PR and adapt based on the other code. The networks are similar.

**Comment by sliwy** [comment]
@bruAristimunha to debug easily the number of parameters change in the layers you can use https://github.com/TylerYep/torchinfo

**Comment by bruAristimunha** [comment]
I refactored it based on the other model, and here is the result of the torchinfo summary. I will go into more depth on this issue, but possibly it will take me a while.

===============================================================
Layer (type:depth-idx)                   Param #
=================================================================
EEGInception                             --
├─ELU: 1-1                               --
├─Ensure4d: 1-2                          --
├─Expression: 1-3                        --
├─_InceptionBlock: 1-4                   --
│    └─ModuleList: 2-1                   --
│    │    └─Sequential: 3-1              6,016
│    │    └─Sequential: 3-2              2,984
│    │    └─Sequential: 3-3              1,492
├─AvgPool2d: 1-5                         --
├─AvgPool2d: 1-6                         --
├─Sequential: 1-7                        --
│    └─CustomPad: 2-2                    --
│    └─Conv2d: 2-3                       2,304
│    └─BatchNorm2d: 2-4                  24
│    └─ELU: 2-5                          --
│    └─AvgPool2d: 2-6                    --
│    └─Dropout: 2-7                      --
│    └─CustomPad: 2-8                    --
│    └─Conv2d: 2-9                       288
│    └─BatchNorm2d: 2-10                 12
│    └─ELU: 2-11                         --
│    └─AvgPool2d: 2-12                   --
│    └─Dropout: 2-13                     --
├─Sequential: 1-8                        --
│    └─Flatten: 2-14                     --
│    └─Linear: 2-15                      3,474
│    └─Softmax: 2-16                     --
=================================================================
Total params: 16,594
Trainable params: 16,594
Non-trainable params: 0
=================================================================
=================================================================
Layer (type:depth-idx)                   Param #
=================================================================
EEGInception                             --
├─ELU: 1-1                               --
├─Ensure4d: 1-2                          --
├─Expression: 1-3                        --
├─_InceptionBlock: 1-4                   --
│    └─ModuleList: 2-1                   --
│    │    └─Sequential: 3-1              6,016
│    │    └─Sequential: 3-2              2,984
│    │    └─Sequential: 3-3              1,492
├─AvgPool2d: 1-5                         --
├─AvgPool2d: 1-6                         --
├─Sequential: 1-7                        --
│    └─CustomPad: 2-2                    --
│    └─Conv2d: 2-3                       2,304
│    └─BatchNorm2d: 2-4                  24
│    └─ELU: 2-5                          --
│    └─AvgPool2d: 2-6                    --
│    └─Dropout: 2-7                      --
│    └─CustomPad: 2-8                    --
│    └─Conv2d: 2-9                       288
│    └─BatchNorm2d: 2-10                 12
│    └─ELU: 2-11                         --
│    └─AvgPool2d: 2-12                   --
│    └─Dropout: 2-13                     --
├─Sequential: 1-8                        --
│    └─Flatten: 2-14                     --
│    └─Linear: 2-15                      3,474
│    └─Softmax: 2-16                     --
=================================================================
Total params: 16,594
Trainable params: 16,594
Non-trainable params: 0
=================================================================

**Comment by sliwy** [comment]
@bruAristimunha I edited your comment to fix formatting. They look the same, one is built using Sequential and the other one just using Module?

File: test/unit_tests/models/test_models.py [hunk 2]
```diff
@@ -120,8 +121,65 @@ def test_eegitnet(input_sizes):
     check_forward_pass(model, input_sizes,)
 
 
-@pytest.mark.parametrize('n_channels,sfreq,n_classes,input_size_s',
-                         [(20, 128, 5, 30), (10, 256, 4, 20), (1, 64, 2, 30)])
+def test_eeginception(input_sizes):
```
Comment (robintibor) [inline test/unit_tests/models/test_models.py:124]: this should be deleted or what?

Comment (bruAristimunha) [reply test/unit_tests/models/test_models.py:124]: Yes, I changed the way of taking the test and I ended up forgetting this old test.

**Comment by cedricrommel** [comment]
> Hello guys, @robintibor and @sliwy,
> 
> I've talked to @cedricrommel, and he's going to take a closer look at this code for the next week.

I found a few problems with the layer dimensions and fix them. Both tests are now passing. For the second test I found in the [authors official code](https://github.com/esantamariavazquez/EEG-Inception/blob/main/EEGInception/EEGInception.py) provided by @bruAristimunha that they set all biases to `False` except in the first inception block. After fixing that and the window size in the test, we could get the number of params announced in the paper.

**Comment by cedricrommel** [comment]
In case it can help:
Capture d’écran 2022-09-15 à 17 36 11

**Review by agramfort — COMMENTED** [review]
@bruAristimunha can you add this model to the doc before merging?

🙏

**Comment by cedricrommel** [comment]
Apparently `math.prod` was introduced after python 3.7 😓 Can you fix that @bruAristimunha?

**Comment by bruAristimunha** [comment]
Done @agramfort and @cedricrommel,

Many thanks, @sliwy and @robintibor, for the review, and thank you, @cedricrommel, for the hard work on this model! I spent a lot of time trying to find these little details, and I am pleased that you, as an expert professional, managed to find them.

Happy to contribute to the library with this model that won last year's competition. I am open for further revisions if you feel it necessary, @agramfort, @robintibor and @sliwy.

**Comment by agramfort** [comment]
looks good https://output.circle-artifacts.com/output/job/a9f2fd77-ee59-4cc3-977e-63b895e24f25/artifacts/0/dev/generated/braindecode.models.EEGInception.html !

🙏 @bruAristimunha @cedricrommel
